rubin_changelog/changelog.py

```diff
@@ -0,0 +1,182 @@
+import concurrent
+import datetime
+import logging
+from concurrent.futures.thread import ThreadPoolExecutor
+from copy import deepcopy
+
+from dateutil.parser import parse
+from sortedcontainers import SortedDict, SortedList
+
+from .eups import EupsData
+from .github import GitHubData
+from .jira import JiraData
+from .rst import Writer
+from .tag import *
+
+log = logging.getLogger("changelog")
+
+
+class ChangeLog:
+    def __init__(self, max_workers: int = 5):
+        self._github_cache = None
+        self._max_workers = max_workers
+
+    @staticmethod
+    def get_package_diff(release: ReleaseType) -> SortedDict:
+        eups = EupsData()
+        eups_data = eups.get_releases(release)
+        result = SortedDict()
+        releases = eups_data['releases']
+        last_release = None
+        for r in releases:
+            if last_release is not None:
+                previous_pkgs = [sub['package'] for sub in last_release]
+                pkgs = [sub['package'] for sub in releases[r]]
+                removed = SortedList(set(previous_pkgs) - set(pkgs))
+                added = SortedList(set(pkgs) - set(previous_pkgs))
+                result[r] = {'added': added, 'removed': removed}
+            last_release = releases[r]
+        return result
+
+    @staticmethod
+    def _fetch(repo: str) -> dict:
+        log.info("Fetching %s", repo)
+        gh = GitHubData()
+        result = dict()
+        pulls = gh.get_pull_requests(repo)
+        tags = gh.get_tags(repo)
+        result["repo"] = repo
+        result["pulls"] = pulls
+        result["tags"] = tags
+        del gh
+        return result
+
+    def _get_package_repos(self, products: SortedList) -> SortedDict:
+        result = SortedDict()
+        result['pulls'] = SortedDict()
+        result["tags"] = SortedDict()
+        gh = GitHubData()
+        repos = gh.get_repos()
+        del gh
+        repo_list = SortedList()
+        for product in products:
+            if product in repos:
+                repo_list.add(product)
+        with ThreadPoolExecutor(
+                max_workers=self._max_workers) as executor:
+            futures = {executor.submit(self._fetch, repo): repo for repo in repo_list}
```

Comment:
I guess there's no way to iterate over a list of repos in the query itself?

Reply:
Yes, that should be possible as well to query several repos in the `gql` query directly